src/main/java/net/neoforged/neoforge/attachment/AttachmentHolder.java

```diff
@@ -24,23 +28,41 @@
  * Implementation class for objects that can hold data attachments.
  * For the user-facing methods, see {@link IAttachmentHolder}.
  */
-public abstract class AttachmentHolder implements IAttachmentHolder {
+public final class AttachmentHolder<T extends IAttachmentHolder> implements IAttachmentHolder {
     public static final String ATTACHMENTS_NBT_KEY = "neoforge:attachments";
-    private static final boolean IN_DEV = !FMLLoader.isProduction();
     private static final Logger LOGGER = LogUtils.getLogger();
 
+    @Nullable
+    final T parent;
+
+    @Nullable
+    Map<AttachmentType<?>, Object> attachments = null;
+
+    public final Codec<AttachmentHolder<T>> CODEC;
+
+    public AttachmentHolder(T parent) {
+        this.parent = parent;
+
+        Class<T> parentClass = (Class<T>) TypeResolver.resolveRawClass(AttachmentHolder.class, parent.getClass());
```

Comment:
Ugh please no. `TypeResolver` is atrocious and ought never to be used. Why exactly is this needed here? And can we, like, design an API so it's not?